skrifa/src/meta/metrics.rs

```diff
@@ -0,0 +1,434 @@
+    if size == 0.0 || upem == 0.0 {
+        1.0
+    } else {
+        size / upem
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::MetadataProvider as _;
+    use read_fonts::test_data::test_fonts::{SIMPLE_GLYF, VAZIRMATN_VAR};
+    use read_fonts::FontRef;
+
+    #[test]
+    fn metrics() {
+        let font = FontRef::new(SIMPLE_GLYF).unwrap();
+        let metrics = font.metrics(None, &[]);
+        let expected = Metrics {
+            units_per_em: 1024,
+            glyph_count: 3,
+            bounds: Some(BoundingBox {
+                x_min: 51.0,
+                y_min: -250.0,
+                x_max: 998.0,
+                y_max: 950.0,
+            }),
+            average_width: Some(1275.0),
+            max_width: None,
+            x_height: Some(512.0),
+            cap_height: Some(717.0),
+            is_monospace: false,
+            italic_angle: 0.0,
+            ascent: 950.0,
+            descent: -250.0,
+            leading: 0.0,
+            underline: None,
+            strikeout: Some(Decoration {
+                offset: 307.0,
+                thickness: 51.0,
+            }),
+        };
+        assert_eq!(metrics, expected);
+    }
+
+    #[test]
+    fn metrics_missing_os2() {
+        let font = FontRef::new(VAZIRMATN_VAR).unwrap();
+        let metrics = font.metrics(None, &[]);
+        let expected = Metrics {
+            units_per_em: 2048,
+            glyph_count: 4,
+            bounds: Some(BoundingBox {
+                x_min: 29.0,
+                y_min: 0.0,
+                x_max: 1310.0,
+                y_max: 1847.0,
+            }),
+            average_width: None,
+            max_width: Some(1336.0),
+            x_height: None,
+            cap_height: None,
+            is_monospace: false,
+            italic_angle: 0.0,
+            ascent: 2100.0,
+            descent: -1100.0,
+            leading: 0.0,
+            underline: None,
+            strikeout: None,
+        };
+        assert_eq!(metrics, expected);
+    }
+
+    #[test]
+    fn glyph_metrics() {
+        let font = FontRef::new(VAZIRMATN_VAR).unwrap();
+        let glyph_metrics = font.glyph_metrics(None, &[]);
+        // (advance_width, lsb) in glyph order
+        let expected = &[
+            (908.0, 100.0),
+            (1336.0, 29.0),
+            (1336.0, 29.0),
+            (633.0, 57.0),
+        ];
+        let result = (0..4)
+            .map(|i| {
+                let gid = GlyphId::new(i as u16);
+                let advance_width = glyph_metrics.advance_width(gid).unwrap();
+                let lsb = glyph_metrics.left_side_bearing(gid).unwrap();
+                (advance_width, lsb)
+            })
+            .collect::<Vec<_>>();
+        assert_eq!(expected, &result[..]);
+    }
+
+    #[test]
+    fn glyph_metrics_var() {
+        let font = FontRef::new(VAZIRMATN_VAR).unwrap();
+        let coords = &[NormalizedCoord::from_f32(-0.8)];
```

Comment:
Could this exhibit the process for *user* coords? - I want to see something bulk up with weight. Something is at -0.8 is a bit impenetrable.

Reply:
I’ll add a test here with user coords in the PR that adds variation metadata so I have the infra in place.